compiler/qsc_rca/src/rca.rs

```diff
@@ -0,0 +1,2340 @@
+        .get(&expr_id)
+        .expect("expression's compute kind should exist");
+    aggregate_compute_kind_runtime_features(ComputeKind::Classical, expr_compute_kind)
+}
+
+fn determine_stmt_semi_compute_kind(
+    package_id: PackageId,
+    expr_id: ExprId,
+    application_instance: &mut ApplicationInstance,
+    package_store: &PackageStore,
+    package_store_scaffolding: &mut PackageStoreScaffolding,
+) -> ComputeKind {
+    // First, analyze the expression.
+    simulate_expr(
+        (package_id, expr_id).into(),
+        application_instance,
+        package_store,
+        package_store_scaffolding,
+    );
+
+    // Use the expression's compute kind to construct the statement's compute kind, only using using the expression's
+    // runtime features since the value is meaningless for semicolon statements.
+    let expr_compute_kind = application_instance
+        .exprs
+        .get(&expr_id)
+        .expect("expression's compute kind should exist");
+    aggregate_compute_kind_runtime_features(ComputeKind::Classical, expr_compute_kind)
+}
+
+fn propagate_applications_table_through_block(
+    block_id: BlockId,
+    applications_table: &ApplicationsTable,
+    package: &impl PackageLookup,
+    package_scaffolding: &mut PackageScaffolding,
+) {
+    let block = package.get_block(block_id);
+    for stmt_id in &block.stmts {
+        propagate_applications_table_through_stmt(
+            *stmt_id,
+            applications_table,
+            package,
+            package_scaffolding,
+        );
+    }
+    package_scaffolding
+        .blocks
+        .insert(block_id, applications_table.clone());
+}
+
+fn map_input_param_exprs(
+    pat_id: StorePatId,
+    expr_id: StoreExprId,
+    package_store: &impl PackageStoreLookup,
+) -> Vec<ExprId> {
+    let pat = package_store.get_pat(pat_id);
+    match &pat.kind {
+        PatKind::Bind(_) | PatKind::Discard => vec![expr_id.expr],
+        PatKind::Tuple(pats) => {
+            let expr = package_store.get_expr(expr_id);
+            match &expr.kind {
+                ExprKind::Tuple(exprs) => {
+                    assert!(pats.len() == exprs.len());
+                    let mut input_param_exprs = Vec::<ExprId>::with_capacity(pats.len());
+                    for (local_pat_id, local_expr_id) in pats.iter().zip(exprs.iter()) {
+                        let global_pat_id = StorePatId::from((pat_id.package, *local_pat_id));
+                        let global_expr_id = StoreExprId::from((expr_id.package, *local_expr_id));
+                        let mut sub_input_param_exprs =
+                            map_input_param_exprs(global_pat_id, global_expr_id, package_store);
+                        input_param_exprs.append(&mut sub_input_param_exprs);
+                    }
+                    input_param_exprs
+                }
+                _ => panic!("expected tuple expression"),
+            }
+        }
+    }
+}
+
+fn propagate_applications_table_through_expr(
```

Comment:
This recursive propagation through exprs, stmts, and blocks feels like it could have been implemented with the visitor pattern... have you tried that?

Reply:
When I started implementing RCA, I tried using the visitor pattern. However, the way I was originally implementing it caused some conflicts with the borrow checker when using the visitor pattern.

I think now that I've implemented RCA, I can refactor it to use the visitor pattern. I'll work on it.